pkgs/development/compilers/tinygo/default.nix

```diff
@@ -1,64 +1,129 @@
-{ lib, buildGoModule, fetchFromGitHub, llvm, clang-unwrapped, lld, avrgcc
-, avrdude, openocd, gcc-arm-embedded, makeWrapper, fetchurl }:
+{ stdenv
+, lib
+, buildGoModule
+, fetchFromGitHub
+, makeWrapper
+, substituteAll
+, llvmPackages
+, libffi
+, wasi-libc
+, avrgcc
+, go
+, binaryen
+, avrdude
+, gdb
+, openocd
+}:
 
-let main = ./main.go;
-    gomod = ./go.mod;
+let
+  llvmMajor = lib.versions.major llvm.version;
+  inherit (llvmPackages) llvm clang compiler-rt lld;
 in
+
 buildGoModule rec {
   pname = "tinygo";
-  version = "0.16.0";
+  version = "0.23.0";
 
   src = fetchFromGitHub {
     owner = "tinygo-org";
-    repo = "tinygo";
+    repo = pname;
     rev = "v${version}";
-    sha256 = "063aszbsnr0myq56kms1slmrfs7m4nmg0zgh2p66lxdsifrfly7j";
+    sha256 = "sha256-YgQGAQJw9Xyw5BF2d9uZTQHfjHsu2evZGo4RV9DtStE=";
     fetchSubmodules = true;
   };
 
-  overrideModAttrs = (_: {
-      patches = [];
-      preBuild = ''
-      rm -rf *
-      cp ${main} main.go
-      cp ${gomod} go.mod
-      chmod +w go.mod
-      '';
-  });
+  patches = [
+    (substituteAll {
+      src = ./0001-Add-clang-header-path.patch;
+      clang_include = "${clang.cc.lib}/lib/clang/${clang.cc.version}/include";
+    })
+    ./0002-Makefile.patch
+  ];
 
-  preBuild = "cp ${gomod} go.mod";
+  vendorSha256 = "sha256-fK8BlCh+1NtHW6MwW68iSIB+Sw6AK+g3y4lMyMYrXkk=";
 
-  postBuild = "make gen-device";
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ llvm libffi clang.cc ];
+  checkInputs = [ avrgcc binaryen ];
 
-  vendorSha256 = "12k2gin0v7aqz5543m12yhifc0xsz26qyqra5l4c68xizvzcvkxb";
+  doCheck = stdenv.buildPlatform == stdenv.hostPlatform;
 
-  doCheck = false;
+  subPackages = [ "." ];
+  tags = [ "llvm${llvmMajor}" ];
+  allowGoReference = true;
 
-  prePatch = ''
-    sed -i s/', "-nostdlibinc"'// builder/builtins.go
-    sed -i s/'"-nostdlibinc", '// compileopts/config.go builder/picolibc.go
-  '';
+  # only strip the compiler
```

Comment:
Why not the other things?

Reply:
Makefile pre-compiles `picolibc` and `compiler-rt` for ARM CPU's which we probably don't want to strip. They are installed at `$out/share/tinygo/pkg/*`.